### PR TITLE
feat: add api spec test infrastructure

### PR DESCRIPTION
**Pull request by tmashuang**
## **Description**

This PR adds tests using [api-specs](https://github.com/MetaMask/api-specs) via the [@open-rpc/test-coverage](https://github.com/open-rpc/test-coverage) tool.

## **Related issues**

Fixes: https://github.com/MetaMask/MetaMask-planning/issues/2238

## **Manual testing steps**

1. `yarn setup`
2. `yarn test:e2e:ios:debug:build`
3. `yarn test:api-specs`

## **Screenshots/Recordings**
Screenshot 2024-06-18 at 3 25 26 PM
image

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by tmashuang** [comment]
I have read the CLA Document and I hereby sign the CLA

File: e2e/specs/window-json-rpc-test/json-rpc-coverage.spec.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+          return result;
+        };
+
+        const createDriverTransport = (driver) => async (_, method, params) => {
+          await driver.runScript(
+            (el, m, p) => {
+              window.ethereum
+                .request({ method: m, params: p })
+                .then((res) => (el.innerHTML = JSON.stringify(res)))
```
Comment (shanejonas) [inline e2e/specs/window-json-rpc-test/json-rpc-coverage.spec.js:68]:
```suggestion
                .then((res) => (el.innerHTML = JSON.stringify({result: res})))
```

File: e2e/specs/window-json-rpc-test/json-rpc-coverage.spec.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                .catch((err) => (el.innerHTML = JSON.stringify(err)));
```
Comment (shanejonas) [inline e2e/specs/window-json-rpc-test/json-rpc-coverage.spec.js:69]:
```suggestion
                .catch((err) => (el.innerHTML = JSON.stringify({error: {code: e.code, message: e.message, data: e.data}})));
```

File: e2e/specs/window-json-rpc-test/json-rpc-coverage.spec.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+          while (result.result === '') {
+            await TestHelpers.delay(500);
+            result = await webElement.runScript(`(el) => {
+              return el.innerHTML
+            }`);
+          }
+
+          // await web.element(by.web.id('json-rpc-response')).replaceText('');
+
+          return result;
```
Comment (shanejonas) [inline e2e/specs/window-json-rpc-test/json-rpc-coverage.spec.js:60]: probably also need to JSON.parse here
```suggestion
          return JSON.parse(result);
```

File: e2e/pages/Settings/NetworksView.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ import Matchers from '../../utils/Matchers';
 import Gestures from '../../utils/Gestures';
 
 class NetworkView {
-  get networkContainer() {
+  get   networkContainer() {
```
Comment (adonesky1) [inline e2e/pages/Settings/NetworksView.js:9]:
```suggestion
  get  networkContainer() {
```

File: index.js [hunk 5]
```diff
@@ -26,55 +26,56 @@ import { isNotificationsFeatureEnabled } from './app/util/notifications/methods'
...
-  'new NativeEventEmitter', // New libraries have not yet implemented native methods to handle warnings (https://stackoverflow.com/questions/69538962/new-nativeeventemitter-was-called-with-a-non-null-argument-without-the-requir)
-  'EventEmitter.removeListener',
-  'Module TcpSockets requires main queue setup',
-  'Module RCTSearchApiManager requires main queue setup',
-  'PushNotificationIOS has been extracted', // RNC PushNotification iOS issue - https://github.com/react-native-push-notification/ios/issues/43
-  "ViewPropTypes will be removed from React Native, along with all other PropTypes. We recommend that you migrate away from PropTypes and switch to a type system like TypeScript. If you need to continue using ViewPropTypes, migrate to the 'deprecated-react-native-prop-types' package.",
-  'ReactImageView: Image source "null"',
-  'Warning: componentWillReceiveProps has been renamed',
-]);
+LogBox.ignoreAllLogs();
```
Comment (shanejonas) [inline index.js:29]: need to revert this part

cc @cortisiko

**Comment by cortisiko** [comment]
@shanejonas @adonesky1 I take it that we intend on including this as part of our test pipeline? If so, we should add the tests to the smoke pipeline. So that you know, our CI platform (bitrise) operates slightly differently than circleCI. 


Here is a visual representation of what our smoke pipeline looks like on bitrise: 
Screen Shot 2024-06-05 at 4 35 45 PM


Now here are the steps we can do for now just as a proof of concept and iterate in the future:

-  we can duplicate this [workflow](https://github.com/MetaMask/metamask-mobile/blob/main/bitrise.yml#L749%23L815) and call it a different name (ios_rpc_coverage_e2e). then change the test [script](https://github.com/MetaMask/metamask-mobile/blob/main/bitrise.yml#L815) to run your tests: 
- we then can create a separate workflow that we will add in the `run_smoke_e2e_ios_android_stage`
  basically, duplicate  https://github.com/MetaMask/metamask-mobile/blob/main/bitrise.yml#L471, call it whatever you want

- make sure the workflow you created above to the  `run_smoke_e2e_ios_android_stage` : https://github.com/MetaMask/metamask-mobile/blob/main/bitrise.yml#L118
- You can rinse and repeat for Android


Feel free to reach out if anything.

File: e2e/api-specs/json-rpc-coverage.js [hunk 6]
```diff
@@ -0,0 +1,494 @@
...
+
+    async afterRequest(_, call) {
+      await new Promise((resolve, reject) => {
+        addToQueue({
+          name: 'afterRequest',
+          resolve,
+          reject,
+          task: async () => {
+            if (call.methodName === 'wallet_addEthereumChain') {
+              const cancelButton = await Matchers.getElementByText('Cancel');
```
Comment (cortisiko) [inline e2e/api-specs/json-rpc-coverage.js:238]: We can use a page object for this. You would need to create the element for the Cancel Button and the user gesture to tap the cancel button. 

You can follow the same steps done with [tapApproveButton](https://github.com/MetaMask/metamask-mobile/blob/e2e-json-rpc-coverage-tool/e2e/pages/modals/NetworkApprovalModal.js#L10).

File: e2e/api-specs/json-rpc-coverage.js [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+
+            if (call.methodName === 'wallet_requestPermissions') {
+              await TestHelpers.delay(3000);
+              const cancelButton = await Matchers.getElementByText('Cancel');
+              await Gestures.tap(cancelButton);
+            }
+
+            if (call.methodName === 'personal_sign') {
+              await TestHelpers.delay(3000);
+              const cancelButton = await Matchers.getElementByText('Cancel');
```
Comment (cortisiko) [inline e2e/api-specs/json-rpc-coverage.js:256]: The SignModal page object have actions for tapping the cancel button.

```suggestion
await SigningModal.tapCancelButton();
```

File: e2e/api-specs/json-rpc-coverage.js [hunk 8]
```diff
@@ -0,0 +1,494 @@
...
+              await Gestures.tap(cancelButton);
+            }
+
+            if (call.methodName === 'eth_signTypedData_v4') {
+              await TestHelpers.delay(3000);
+              const cancelButton = await Matchers.getElementByText('Cancel');
```
Comment (cortisiko) [inline e2e/api-specs/json-rpc-coverage.js:262]: The SignModal page object have actions for tapping the cancel button.

```suggestion
await SigningModal.tapCancelButton();
```

File: e2e/api-specs/json-rpc-coverage.js [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+
+            if (call.methodName === 'wallet_switchEthereumChain') {
+              await TestHelpers.delay(3000);
+              const cancelButton = await Matchers.getElementByText('Cancel');
+              await Gestures.tap(cancelButton);
+            }
+
+            if (call.methodName === 'wallet_requestPermissions') {
+              await TestHelpers.delay(3000);
+              const cancelButton = await Matchers.getElementByText('Cancel');
```
Comment (cortisiko) [inline e2e/api-specs/json-rpc-coverage.js:250]: The ConnectModal page object have actions for tapping the cancel button.
```suggestion
              await ConnectModal.tapCancelButton();
```

File: e2e/api-specs/json-rpc-coverage.js [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+      ganacheOptions: defaultGanacheOptions,
+      disableGanache: true,
+      restartDevice: true,
+    },
+    async () => {
+      await loginToApp();
+      await TabBarComponent.tapBrowser();
+      await Browser.navigateToTestDApp();
+
+      const webElement = await web.element(by.web.id('json-rpc-response'));
```
Comment (cortisiko) [inline e2e/api-specs/json-rpc-coverage.js:328]: Create an element for this button in the TestDapp page object. Follow the exact steps [here](https://github.com/MetaMask/metamask-mobile/blob/e2e-json-rpc-coverage-tool/e2e/pages/Browser/TestDApp.js#L25) to create the element.

File: e2e/api-specs/json-rpc-coverage.js [hunk 11]
```diff
@@ -0,0 +1,494 @@
...
+      await webElement.scrollToView();
```
Comment (cortisiko) [inline e2e/api-specs/json-rpc-coverage.js:329]: Then in the TestDapp page object you can have a method that scrolls to the button. Something like scrollToJsonRPCResultSection()

[here](https://github.com/MetaMask/metamask-mobile/blob/e2e-json-rpc-coverage-tool/e2e/pages/Browser/TestDApp.js#L122) is how we scroll to a webview element

**Review by cortisiko — CHANGES_REQUESTED** [review]
good stuff @shanejonas I left a few comments.  Sharing https://github.com/MetaMask/contributor-docs/blob/main/docs/e2e/mobile-e2e-guidelines.md#step-2-create-the-networks-page-object in case you want a quick run down on how we create page objects within our tests!

File: e2e/api-specs/json-rpc-coverage.js (hunk 11 above)
Comment (shanejonas) [reply e2e/api-specs/json-rpc-coverage.js:329]: This `json-rpc-response` element (or scrolling to it) isnt actually needed since I refactored the existing code that relied on the test-dapp.

File: e2e/api-specs/json-rpc-coverage.js (hunk 10 above)
Comment (shanejonas) [reply e2e/api-specs/json-rpc-coverage.js:328]: This `json-rpc-response` element isn't actually needed. but going to change it to `tag('body')`

File: package.json [hunk 12]
```diff
@@ -56,6 +56,7 @@
     "test": "yarn test:unit && yarn test:e2e",
     "test:unit": "jest ./app/ ./locales/",
     "test:unit:update": "time jest -u ./app/",
+    "test:api-specs": "detox test -c ios.sim.apiSpecs",
```
Comment (jiexi) [inline package.json:59]:
```suggestion
    "test:api-specs": "detox build -c ios.sim.release && detox test -c ios.sim.apiSpecs",
```

Comment (jiexi) [reply package.json:59]: getting this error when running things for the first time using just `detox test -c ios.sim.apiSpecs`
```
Failed to find the app binary at:
/Users/jiexi/Projects/metamask-mobile/ios/build/Build/Products/Debug-iphonesimulator/MetaMask.app

```

**Comment by Cal-L** [comment]
@shanejonas nice work on this! Just curious, were you ever able to get the mock server working?

**Comment by shanejonas** [comment]
@Cal-L mock server stuff is working now, but an unrelated android e2e suite is failing.

File: bitrise.yml [hunk 13]
```diff
@@ -1400,6 +1470,8 @@ meta:
     stack: osx-xcode-15.0.x
     machine_type_id: g2-m1-max.5core
 trigger_map:
+  - push_branch: e2e-json-rpc-coverage-tool
```
Comment (cortisiko) [inline bitrise.yml:1473]: remember to remove this prior to merge.

**Comment by shanejonas** [comment]
@SocketSecurity ignore npm/json-schema-ref-parser@6.1.0
@SocketSecurity ignore npm/@mantine/hooks@7.8.0
@SocketSecurity ignore npm/@open-rpc/server-js@1.9.3
@SocketSecurity ignore npm/@open-rpc/test-coverage@2.2.2

**Review by cortisiko — DISMISSED** [review]
🌮 🌮 🌮  Looks good!

File: e2e/api-specs/json-rpc-coverage.js [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+import { web } from 'detox';
+import detox from 'detox/internals';
+import rpcCoverageTool from '@open-rpc/test-coverage';
+import { parseOpenRPCDocument } from '@open-rpc/schema-utils-js';
+import JsonSchemaFakerRule from '@open-rpc/test-coverage/build/rules/json-schema-faker-rule';
+import HtmlReporter from '@open-rpc/test-coverage/build/reporters/html-reporter';
+
+import Browser from '../pages/Browser/BrowserView';
+// eslint-disable-next-line import/no-commonjs
+const mockServer = require('@open-rpc/mock-server/build/index').default;
```
Comment (tommasini) [inline e2e/api-specs/json-rpc-coverage.js:11]: Any reason we can't use import here?

**Review by tommasini — DISMISSED** [review]
LGTM

**Review by cortisiko — APPROVED** [review]
LGTM!

File: e2e/api-specs/json-rpc-coverage.js (hunk 14 above)
Comment (shanejonas) [reply e2e/api-specs/json-rpc-coverage.js:11]: we can change this later i dont think we should block on this

**Review by tommasini — APPROVED** [review]
